feat: Реализован CustomLinkedList и тесты

Colgens opened the pull request:
- Реализовал CustomLinkedList с методами: add, get, remove, contains, addAll, size, clear, isEmpty, toString.
- Добавлены комплексные модульные тесты для CustomLinkedList.
- В тестах реализовано преобразование потока в CustomLinkedList с использованием reduce.

Colgens committed:
feat: Реализован CustomLinkedList и тесты

- Реализовал CustomLinkedList с методами: add, get, remove, contains, addAll, size, clear, isEmpty, toString.
- Добавлены комплексные модульные тесты для CustomLinkedList.
- Реализовано преобразование потока в CustomLinkedList с использованием reduce.

kraskden commented on src/main/java/com/tbank/fintech_project/collections/CustomLinkedList.java:
```diff
@@ -0,0 +1,187 @@
+        }
+    }
+
+    private ListNode<T> getNode(int index) {
```
красиво

kraskden commented on src/main/java/com/tbank/fintech_project/collections/CustomLinkedList.java:
```diff
@@ -0,0 +1,187 @@
+    }
+
+    public void addAll(CustomLinkedList<T> elements) {
+        for (int i = 0; i < elements.size; i++) {
```
Плохая реализация с O(n^2), т.к .get(i) работает за O(N) и вызывается в цикле N раз

Склеивание листов должно происходить за O(1), там нужно просто поиграться с ссылочками prev/next

Colgens replied:
> Плохая реализация с O(n^2), т.к .get(i) работает за O(N) и возникает в цикле N раз
> 
> Склеивание файлов должно происходить за O(1), там нужно просто поиграться с ссылочками предыдущая/следующая

Всё исправил, теперь O(1)

Colgens committed:
refactor: Улучшена производительность addAll и добавлены методы вставки по индексу

- Исправлен метод addAll для работы за O(1) времени путем манипуляции ссылками prev/next.
- Добавлен метод addAll(int index, CustomLinkedList<T> elements) для вставки LinkedList по указанному индексу.
- Добавлен метод addAll(int index, Collection<? extends T> elements) для вставки коллекции по указанному индексу.
- Создан приватный метод addAllInternal для обработки общей логики добавления элементов по указанному индексу.
